### ISSUE TITLE
File Uploads very slow (stuck at 0%)

### ISSUE THREAD

**Issue by tobiasreinhard** [RocketChat/Rocket.Chat#24225]
### Description:

Since the migration to MongoDB 5.0 with "wiredTiger", the file uploads in all channels are very very slow. If you upload a file or a screenshot, the progress bar gets stuck at 0%. After 1 Minute, the file is uploaded, but the preview image can't be opened. After more than 30 seconds, the preview may open.

### Steps to reproduce:

1. Upgrade RC to 4.3.1 (Docker) with MongoDB 5.0.5 (Docker)
2. Try uploading a file to a channel or private conversation
3. Watch your file being uploaded, but the progress bar is stuck at 0%

### Expected behavior:

Instantly upload the file and show the preview if you click on them.

### Actual behavior:

Uploading stucks at 0%, file gets uploaded very very slow and the preview won't show for more than 1 minute.

### Server Setup Information:

- Version of Rocket.Chat Server: 4.3.1
- Operating System: Debian 10
- Deployment Method: Docker (with ReverseProxy)
- Number of Running Instances: 1
- DB Replicaset Oplog: 252
- NodeJS Version: 12.22.8
- MongoDB Version: 5.0.5 / wiredTiger (oplog activated)
- Used file storage: GridFS

### Client Setup Information

- Desktop App or Browser Version:  3.7.6
- Operating System: Windows 10

### Relevant logs:
Can't see relevant entrys in both SERVER and CLIENT logs.

**Comment by tobiasreinhard** [comment]
![grafik](https://user-images.githubusercontent.com/10668507/150317400-444103c1-48cc-477e-a12f-d48ac80e2175.png)

Additionally, the "Person X is uploading..." message won't dissapear.

**Comment by tobiasreinhard** [comment]
Found this in the Nginx Error Log:

`
2022/01/20 11:42:13 [error] 26214#26214: *415 upstream timed out (110: Connection timed out) while reading response header from upstream, client: 192.168.20.4, server: chat.domain.xyz, request: "POST /api/v1/rooms.upload/QD5v4Xp7F3uySNhba HTTP/2.0", upstream: "http://127.0.0.1:3000/api/v1/rooms.upload/QD5v4Xp7F3uySNhba", host: "chat.domain.xyz", referrer: "https://chat.domain.xyz/direct/QD5v4Xp7F3uySNhba"
`

**Comment by zwnk** [comment]
i had a similiar issue after updating to 5.0.3. It was solved by moving the db, data folder to an ssd.

**Comment by zwnk** [comment]
i saw you are using the latest rc version with mongodb 5 and docker. Could you please send me your docker-compose file because i'm having an issue updating it.

**Comment by tobiasreinhard** [comment]
Sure, this is our latest docker-compose file.

```
version: '2'

services:
  rocketchat:
    image: rocketchat/rocket.chat:latest
    command: >
      bash -c
        "for (( ; ; )); do
          node main.js &&
          s=$$? && break || s=$$?;
          echo \"Could not start Rocket.Chat. Waiting 5 secs...\";
          sleep 5;
        done; (exit $$s)"
    restart: unless-stopped
    volumes:
      - ./uploads:/app/uploads
    environment:
      - PORT=3000
      - ROOT_URL=https://chat.domain.tld
      - MONGO_URL=mongodb://mongo:27017/rocketchat
      - MONGO_OPLOG_URL=mongodb://mongo:27017/local
      - MAIL_URL=smtp://smtp.email
    depends_on:
      - mongo
    ports:
      - 3000:3000
    labels:
      - "traefik.backend=rocketchat"
      - "traefik.frontend.rule=Host: your.domain.tld"

  mongo:
    image: mongo:5.0
    restart: unless-stopped
    volumes:
      - ./data/db:/data/db
      - ./data/dump:/dump
    command: >
      bash -c
        "while [ ! -f /data/db/WiredTiger ]; do
          echo \"wiredTiger migration hasn't started yet. Waiting 30 secs...\";
          sleep 30;
        done;
        docker-entrypoint.sh mongod --oplogSize 128 --replSet rs0 --storageEngine=wiredTiger;"
    depends_on:
      - migrator
    labels:
      - "traefik.enable=false"
    ports:
      - 27017:27017

  migrator:
    build: ./docker/
    volumes:
      - ./data/db:/data/db

  mongo-init-replica:
    image: mongo:4.4
    command: >
      bash -c
        "for (( ; ; )); do
          mongo mongo/rocketchat --eval \"
            rs.initiate({
              _id: 'rs0',
              members: [ { _id: 0, host: 'localhost:27017' } ]})\" &&
          s=$$? && break || s=$$?;
          echo \"Could not reach MongoDB. Waiting 5 secs ...\";
          sleep 5;
        done; (exit $$s)"
    depends_on:
      - mongo
```

**Comment by tobiasreinhard** [comment]
Still the same problem, any news on this?

**Comment by IgorOhrimenko** [comment]
I try reproduce issue on the new installation and issue does not reproduce.
https://user-images.githubusercontent.com/24807060/160808790-50194467-feb6-41d9-83c1-ba06d917cfb0.mp4
@tobiasreinhard, your problem reproduce is only on the old big database after migrate?

**Comment by tobiasreinhard** [comment]
> 

Hello Igor, yes, only in the old big database.

**Comment by artemChernitsov** [comment]
I had the same issue
Looks like after database migration by RocketChat manual all forget about indexes of uploads.
It is solved by this comment https://github.com/RocketChat/Rocket.Chat/issues/23467#issuecomment-957126927

Standalone setup
```
mongo
use rocketchat
db.rocketchat_uploads.chunks.createIndex( { files_id: 1, n: 1 }, { unique: true } )
```

Docker setup 
```
docker exec -it mongodb mongo
use rocketchat
db.rocketchat_uploads.chunks.createIndex( { files_id: 1, n: 1 }, { unique: true } )
```

**Comment by tobiasreinhard** [comment]
I tried your solution on our RC instance, switched back from FileSystem to GridFS, but the file uploads are still very slow and sometimes them seem to be stuck.

**Comment by vmsman** [comment]
I tried the database re-index and my Rocketchat uploads freeze at 0% desktop or mobile.  I migrated from mmap to wiredTiger and I am at mongo 5 now.  Everything else works except uploads.  I have tried filesystem and Griff's and no difference.  I have always used Griff's and it worked until the end migration using the process on the Rocketchat website.

**Comment by v360-eduardo-marques** [comment]
facing the same problem here. did someone was able to fix it?

**Comment by vmsman** [comment]
Yes. I was able to fix it.  Surprisingly, my problem was that I was using NginX Proxy Manager (NPM) as my front end network reverse proxy.  There was apparently some corruption in it.  It was REALLY SUBTLE.  I ended up reinstalling NPM and the RocketChat uploads worked again.  The thing that clued me was that my discourse forum uploads stopped working too with a 502 error.  After reinstalling NPM, RocketChat and Discourse uploads worked again.

**Comment by tobiasreinhard** [comment]
I've still not fixed the issue. We ended up using the Filesystem as storage. Sometimes files are completly missing, but we can't afford time to reinstall the complete chat.

**Comment by ncedomir** [comment]
This is not working for me

docker exec -it mongodb mongo
use rocketchat
db.rocketchat_uploads.chunks.createIndex( { files_id: 1, n: 1 }, { unique: true } )

I use Docker & I have the following

+--------------------------------------------------+
|                  SERVER RUNNING                  |
+--------------------------------------------------+
|                                                  |
|  Rocket.Chat Version: 6.0.0                      |
|       NodeJS Version: 14.21.2 - x64              |
|      MongoDB Version: 5.0.15                     |
|       MongoDB Engine: wiredTiger                 |
|             Platform: linux                      |
|         Process Port: 3005                       |
|             Site URL: https://chat.example.com  |
|     ReplicaSet OpLog: Enabled                    |
|          Commit Hash: aa8f0ec70b                 |
|        Commit Branch: HEAD                       |
|                                                  |

Should I go above MongoDB5 5 ?

**Comment by arashkhavari** [comment]
> > I had the same issue Looks like after database migration by RocketChat manual all forget about indexes of uploads. It is solved by this comment [#23467 (comment)](https://github.com/RocketChat/Rocket.Chat/issues/23467#issuecomment-957126927)
> > Standalone setup
> > ```
> > mongo
> > use rocketchat
> > db.rocketchat_uploads.chunks.createIndex( { files_id: 1, n: 1 }, { unique: true } )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Docker setup
> > ```
> > docker exec -it mongodb mongo
> > use rocketchat
> > db.rocketchat_uploads.chunks.createIndex( { files_id: 1, n: 1 }, { unique: true } )
> > ```
> 
> any way to do this with a snap installation?

Install just mongosh client:
[https://www.mongodb.com/docs/mongodb-shell/install/](https://www.mongodb.com/docs/mongodb-shell/install/)
